test/FastTests/Corax/OrderByMultiSorting.cs

```diff
@@ -0,0 +1,181 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Corax;
+using Corax.Queries;
+using FastTests.Voron;
+using Sparrow.Server;
+using Sparrow.Threading;
+using Voron;
+using Xunit;
+using Xunit.Abstractions;
+using static Corax.Queries.SortingMatch;
+
+namespace FastTests.Corax
+{
+
+    public class OrderByMultiSortingTests : StorageTest
+    {
+        private List<IndexSingleNumericalEntry<long, long>> longList = new();
+        private IndexSearcher _indexSearcher;
+        private const int IndexId = 0, Content1 = 1, Content2 = 2;
+
+        public OrderByMultiSortingTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public void OrderByNoRepetitions()
+        {
+            PrepareData();
+
+            IndexEntries();
+            longList.Sort(CompareDescending);
+            using var searcher = new IndexSearcher(Env);
+            {
+                var match1 = searcher.AllEntries();
+
+                var comparer1 = new DescendingMatchComparer(searcher, Content1, MatchCompareFieldType.Integer);
+                var comparer2 = new AscendingMatchComparer(searcher, Content2, MatchCompareFieldType.Integer);
+                var match = SortingMultiMatch.Create(searcher, match1, comparer1, comparer2);
+
+                List<string> sortedByCorax = new();
+                Span<long> ids = stackalloc long[2048];
+                int read = 0;
+                do
+                {
+                    read = match.Fill(ids);
+                    for (int i = 0; i < read; ++i)
+                        sortedByCorax.Add(searcher.GetIdentityFor(ids[i]));
+                }
+                while (read != 0);
+
+                for (int i = 0; i < longList.Count; ++i)
+                    Assert.Equal(longList[i].Id, sortedByCorax[i]);
+
+                Assert.Equal(1000, sortedByCorax.Count);
+            }
+        }
+
+        [Fact]
+        public void OrderByWithRepetitions()
+        {
+            PrepareData();
+            PrepareData(inverse:true);
+
+            IndexEntries();
+            longList.Sort(CompareAscendingThenDescending);
+            using var searcher = new IndexSearcher(Env);
+            {
+                var match1 = searcher.AllEntries();
+
+                var comparer1 = new AscendingMatchComparer(searcher, Content1, MatchCompareFieldType.Integer);
+                var comparer2 = new DescendingMatchComparer(searcher, Content2, MatchCompareFieldType.Integer);
+               
+                var match = SortingMultiMatch.Create(searcher, match1, comparer1, comparer2);
```

Comment:
This is how to use it @maciejaszyk

Reply:
The question is how to do that from an API that doesn't know about this upfront?

Reply:
Everything here is constructed in runtime, what do you mean?

Reply:
I'm giving you a `List<string> OrderBy`, what do you do? This is strong typed on the number of sorts, no?

Reply:
Yes, I talked with @maciejaszyk about it. Yes that is correct, however we arrived at a tradeoff that would be good enough and will give us a lot of leeway to change the implementation over time. We will relax the requirement for the type to be a `struct` which will give us 2 things:
- In those cases that we cannot know the list of comparer types, we just use `IMatchComparer` instead. 
- We can actually optimize internally all those usages with the front facing API never know that such thing would happen.
- In case of custom comparers we can start with basic interface calls, and then if we need to we just can emit the code to deal with those too if we find edge cases where it matters.

Reply:
Note that specializing for 2,3 & 4 sort orders should probably suffice for 99.999% of cases, yep